File: go/api/jdy/file.go
/**
 * Copyright (c) 2015-2022, FineX, All Rights Reserved.
 * @author Thomas.Zhuang
 * @date 2022/5/17
 */

package jdy

import (
	"api-demo/base"
	"api-demo/model"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"os"
)

// FileApiClient 文件Api
type FileApiClient struct {
	*base.ApiClient
}

func (api *FileApiClient) ValidVersion(version string) bool {
	return map[string]bool{
		"v1": true,
	}[version]
}

func (api *FileApiClient) DefaultVersion() string {
	return "v1"
}

// DoRequest 发送http请求
func (api *FileApiClient) DoRequest(option *base.RequestOption) (responseBody []byte, err error) {
	if !api.ValidVersion(api.Version) {
		api.Version = api.DefaultVersion()
	}
	return api.ApiClient.DoRequest(option)
}

// UploadToken 获取文件上传凭证和上传地址接口
func (api *FileApiClient) UploadToken(appId, entryId, transactionId string) (responseBody *model.TokenAndURLList, err error) {
	payload, err := json.Marshal(map[string]interface{}{
		"transaction_id": transactionId,
	})
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    fmt.Sprintf("app/%s/entry/%s/file/get_upload_token", appId, entryId),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UploadFile 文件上传接口
func (api *FileApiClient) UploadFile(url, token string, file *os.File) (responseBody *model.FileKey, err error) {
	body := new(bytes.Buffer)
	writer := multipart.NewWriter(body)

	writer.WriteField("token", token)
	w, err := writer.CreateFormFile("file", file.Name())
	if err != nil {
		return
	}
	_, err = io.Copy(w, file)
	if err != nil {
		return
	}
	writer.Close()

	req, err := http.NewRequest(http.MethodPost, url, body)
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", writer.FormDataContentType())

	response, err := http.DefaultClient.Do(req)
	if err != nil {
		return
	}
	defer response.Body.Close()

	r, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return
	}

	if response.StatusCode != http.StatusOK {
		err = fmt.Errorf("request fail, response code: %d, response body: %s", response.StatusCode, string(r))
		return
	}

	err = json.Unmarshal(r, &responseBody)
	return
}

File: python/src/api/jdy/file.py
from ...constants.http_constant import HttpConstant
from ...model.http.http_request_param import HttpRequestParam
from ...model.http.http_request_param import generateTransactionId
from src.model.http.api_client import ApiClient

# 合法的版本
valid_versions = ('v5')
# 默认版本
default_version = 'v5'


class FileApiClient(ApiClient):

    def __init__(self, api_key, host):
        ApiClient.__init__(self, api_key, host, valid_versions, default_version)

    """
    获取文件上传凭证和上传地址接口
    Arguments:  
        work_flow_query_param: WorkFlowApprovalCommentQueryParam 实例
    """

    def uploadToken(self, app_id, entry_id, version=default_version):
        path = HttpConstant.FORM_PATH.format(suffix='file/get_upload_token',
                                             version=self.getValidVersion(version))
        request_param = HttpRequestParam(path, {'transaction_id': generateTransactionId(),
                                                'app_id': app_id,
                                                'entry_id': entry_id})
        return self.send_post(request_param)

    """
    上传文件
    Arguments:
        url: str 
        token: str
        file: 
    """

    def uploadFile(self, url, token, file):
        data = {'token': token}
        return self.send_post_with_file(url, data, file)
